internal/graph/memory.go

```diff
@@ -133,22 +155,39 @@ func (m *Memory) index(ctx context.Context, fetcher content.Fetcher, node ocispe
 	defer m.lock.Unlock()
 
 	// index the node
-	nodeKey := descriptor.FromOCI(node)
+	nodeKey := node.Digest
 	m.nodes[nodeKey] = node
 
 	// for each successor, put it into the node's successors list, and
 	// put node into the succeesor's predecessors list
-	successorSet := set.New[descriptor.Descriptor]()
+	successorSet := set.New[digest.Digest]()
 	m.successors[nodeKey] = successorSet
 	for _, successor := range successors {
-		successorKey := descriptor.FromOCI(successor)
+		successorKey := successor.Digest
 		successorSet.Add(successorKey)
 		predecessorSet, exists := m.predecessors[successorKey]
 		if !exists {
-			predecessorSet = set.New[descriptor.Descriptor]()
+			predecessorSet = set.New[digest.Digest]()
 			m.predecessors[successorKey] = predecessorSet
 		}
 		predecessorSet.Add(nodeKey)
 	}
 	return successors, nil
 }
+
+// Exists returns if a blob denoted by its digest exists in the memory.
+func (m *Memory) Exists(dgst digest.Digest) bool {
+	_, exists := m.nodes[dgst]
+	return exists
+}
+
+// GetDanglingLayers returns the dangling (unreferenced) layer nodes in the memory.
+func (m *Memory) GetDanglingLayers() []ocispec.Descriptor {
```

Comment:
Wondering if this should be called `GetUnreferencedNodes`? The returned nodes can be either a manifest or a layer blob, right?

Reply:
Renamed the function to `GetUnreferencedRootNodes`.